src/ImageSharp/Formats/Qoi/QoiDecoderCore.cs

```diff
@@ -0,0 +1,287 @@
+// Copyright (c) Six Labors.
+// Licensed under the Six Labors Split License.
+
+using System.Buffers.Binary;
+using System.Diagnostics.CodeAnalysis;
+using SixLabors.ImageSharp.IO;
+using SixLabors.ImageSharp.Memory;
+using SixLabors.ImageSharp.Metadata;
+using SixLabors.ImageSharp.PixelFormats;
+
+namespace SixLabors.ImageSharp.Formats.Qoi;
+
+internal class QoiDecoderCore : IImageDecoderInternals
+{
+    /// <summary>
+    ///     The global configuration.
+    /// </summary>
+    private readonly Configuration configuration;
+
+    /// <summary>
+    ///     Used the manage memory allocations.
+    /// </summary>
+    private readonly MemoryAllocator memoryAllocator;
+
+    /// <summary>
+    ///     Gets or sets a value indicating whether the metadata should be ignored when the image is being decoded.
+    /// </summary>
+    private readonly bool skipMetadata;
+
+    /// <summary>
+    ///     The QOI header.
+    /// </summary>
+    private QoiHeader header;
+
+    public QoiDecoderCore(DecoderOptions options)
+    {
+        this.Options = options;
+        this.configuration = options.Configuration;
+        this.skipMetadata = options.SkipMetadata;
+        this.memoryAllocator = this.configuration.MemoryAllocator;
+    }
+
+    public DecoderOptions Options { get; }
+
+    public Size Dimensions { get; }
+
+    /// <inheritdoc />
+    public Image<TPixel> Decode<TPixel>(BufferedReadStream stream, CancellationToken cancellationToken)
+        where TPixel : unmanaged, IPixel<TPixel>
+    {
+        // Process the header to get metadata
+        this.ProcessHeader(stream);
+
+        // Create Image object
+        ImageMetadata metadata = new()
+        {
+            DecodedImageFormat = QoiFormat.Instance,
+            HorizontalResolution = this.header.Width,
+            VerticalResolution = this.header.Height,
+            ResolutionUnits = PixelResolutionUnit.AspectRatio
+        };
+        Image<TPixel> image = new(this.configuration, (int)this.header.Width, (int)this.header.Height, metadata);
+        Buffer2D<TPixel> pixels = image.GetRootFramePixelBuffer();
+
+        this.ProcessPixels(stream, pixels);
+
+        return image;
+    }
+
+    /// <inheritdoc />
+    public ImageInfo Identify(BufferedReadStream stream, CancellationToken cancellationToken)
+    {
+        ImageMetadata metadata = new();
```

Comment:
If we can efficiently parse out `QoiChannels` and `QoiColorSpace` without reading the pixel data, this should be `QoiMetadata`, and we should do the metadata parsing here. Tests also need to be updated to validate it works.

Reply:
I'm sorry I haven't done this, I don't understand what do I have to do

Do I create an `ImageMetadata` in `QoiDecoder`? Or an `QoiMetadata`?
Or `QoiChannels` and `QoiColorSpace` variables in the `QoiDecoder`?

Reply:
You would create and assign `QoiMetadata` in your decoder then add that to the `ImageMetadata`

Reply:
Sorry, for some reason I thought `QoiMetadata` is a subclass of `ImageMetaData`, so my comment was super confusing. 

This is how it actually works:
`ImageMetaData` is a container of format-specific metadata objects. The decoder is expected to fill the metadata object ( in your case `QoiMetadata`) belonging to the format in both `Decode` and `Identify`. You can ask `ImageMetaData.GetQoiMetadata()` to "get or create" your format-specific metadata object which you can then fill from the the data you parse out to `QoiHeader`. Here's how PNG creates its' metadata object:

https://github.com/SixLabors/ImageSharp/blob/2738b3ab5793bddb96f3a5317d46afa0529220ce/src/ImageSharp/Formats/Png/PngDecoderCore.cs#L147-L148